[Core][Model] Add simple_model_runner and a new model XLMRobertaForSequenceClassification through multimodal interface

AllenDou opened the pull request:
This PR, 
1. Add a new model XLMRobertaForSequenceClassification (for RAG scenario) [https://github.com/huggingface/transformers/blob/v4.42.3/src/transformers/models/xlm_roberta/modeling_xlm_roberta.py](url)
which processes input data through a multimodal interface like the following
```
outputs = llm.process([{
    "prompt": prompt,
    "multi_modal_data": {
        "xlmroberta": inputs,
    }
}])
```
2. Introduce `class ModelMode [DECODER, ENCODER, ENCODER_DECODER, EMBEDDING, SIMPLE]`
3. Add a new model runner(simple_model_runner.py), I didn't use model_runner.py or embedding_model_runner.py because they have procedures like logits(), sample(), and pooling() that we don't need. Maybe we could find a way to create a more general model runner to support 'any' model
4. Add SimpleSequenceGroupOutput, SimpleRequestOutput, and SimpleOutput for general output
5. Tensor-parallel is enabled, but no benefits have been achieved so far
6. Test case/example added

I have two goals:
1. To find a way to enable vllm to integrate 'any' model
2. To replace these models' layers with vllm/layers/* as well as TP/PP to achieve performance improvement

CLOSE #6424
CLOSE #6789
CLOSE #8022

DarkLight1337 commented:
If I'm understanding this PR correctly, you are basically using the multi-modal interface to pass data directly to the model (in this case the input IDs and attention mask). Are you working towards making vLLM function out-of-the-box with generic HuggingFace models?

DarkLight1337 commented on examples/flagembedding.py:
```diff
@@ -0,0 +1,6 @@
+reranker = FlagReranker('BAAI/bge-reranker-base', use_fp16=True)
+score = reranker.compute_score([("hello world", "nice to meet you"),
+                                ("head north", "head south")])
+print(score)
```
I don't quite see the value of this example in vLLM. Are you planning to update it to use vLLM components?

AllenDou replied:
In RAG scenario, user usually use FlagEmbedding package, I left it as an example, but could be removed.

DarkLight1337 commented on tests/conftest.py:
```diff
@@ -417,6 +421,17 @@ def generate_greedy_logprobs_limit(
     def encode(self, prompts: List[str]) -> List[List[torch.Tensor]]:
         return self.model.encode(prompts)
 
+    def encode_simple(
```
Imo the naming is a bit misleading since you're returning the logits of the classification head, not the encoded input directly.

AllenDou replied:
Yes, you are right, I use encode_* because I did't find a proper func name for this general 'any' model scenario, maybe someone could give me some suggestion.

DarkLight1337 replied:
Hmm I think since this is so general, you can extract the logits inside your specific test rather than in `conftest.py`. I'm tempted to just call it `forward` since all it does is pass the inputs directly to the model.

AllenDou replied:
Good idea, now it's late, I'll refactor it tomorrow, thank for your advice :)

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -422,10 +423,10 @@ def encode(
-            raise ValueError(
-                "LLM.encode() is only supported for embedding models (XModel)."
-            )
+        if not self.llm_engine.model_config.embedding_mode and \
```
Same here, it's not really "encoding" the input.

DarkLight1337 reviewed (commented):
Left some initial comments.

AllenDou commented:
> If I'm understanding this PR correctly, you are basically using the multi-modal interface to pass data directly to the model (in this case the input IDs and attention mask). Are you working towards making vLLM function out-of-the-box with generic HuggingFace models?

Yes, you are right. In fact, I have two goals:
1. To find a way to enable vllm to integrate 'any' model.
2. To replace these models' layers with vllm/layers/* as well as TP/PP to achieve performance improvement.

After I replaced XLMRobertaForSequenceClassification's query/key/value linear with QKVParallelLinear, I saw a 15% performance improvement.

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -281,8 +282,7 @@ def generate(
         """
-        if self.llm_engine.model_config.embedding_mode or \
-            self.llm_engine.model_config.simple_mode:
+        if self.llm_engine.model_config.embedding_mode:
```
We should still check `simple_mode` here imo. Perhaps we should make the mode an enum to avoid complex if-else cases as we add more modes..

AllenDou replied:
Good idea, Let me try.

AllenDou replied:
I created a class ModelMode to indicate the model mode [ENCODER, DECODER, EMBEDDING, SIMPLE]. Please take a look and give me some suggestions :)

DarkLight1337 replied:
Added a new round of comments!

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -451,6 +451,73 @@ def encode(
+
+        This class automatically batches the given prompts, considering
+        the memory constraint. For the best performance, put all of your prompts
+        into a single list and pass it to this method.
```
Should update the docstring here.

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -451,6 +451,73 @@ def encode(
+            A list of `SimpleRequestOutput` objects containing the
+            generated simple result in the same order as the input prompts.
+
+        Note:
+            Using ``prompts`` and ``prompt_token_ids`` as keyword parameters is
+            considered legacy and may be deprecated in the future. You should
+            instead pass them via the ``inputs`` parameter.
```
To simplify the code, it may be a good idea to drop the old API for this new method.

AllenDou replied:
Sorry, I didn't get your point, the 'old API' mean?  prompt_token_ids? prompts? these signature?

DarkLight1337 replied:
Yeah, that's what I mean. We can have this function only accept `inputs`.

DarkLight1337 commented on vllm/config.py:
```diff
@@ -175,6 +190,7 @@ def __init__(
         if not self.skip_tokenizer_init:
             self._verify_tokenizer_mode()
         self._verify_embedding_mode()
+        self._verify_simple_mode()
```
Let's combine these two functions together and move the logic of determining model mode to model registry. (May have to move `ModelMode` as well to avoid circular imports)

DarkLight1337 commented on vllm/engine/llm_engine.py:
```diff
@@ -259,7 +260,9 @@ def __init__(
         )
 
-        if not self.model_config.embedding_mode:
+        if self.model_config.model_mode not in [
+                ModelMode.EMBEDDING, ModelMode.SIMPLE
+        ]:
```
Perhaps we should let each mode declare whether to set up KV cache instead of handling each case inside `LLMEngine`.

DarkLight1337 commented on vllm/core/scheduler.py:
```diff
@@ -279,8 +279,10 @@ def __init__(
         if self.scheduler_config.use_v2_block_manager:
             version = "v2"
-        if self.scheduler_config.embedding_mode:
+        if self.scheduler_config.model_mode == ModelMode.EMBEDDING:
             version = "embedding"
+        if self.scheduler_config.model_mode == ModelMode.SIMPLE:
+            version = "simple"
```
Perhaps we should let each mode override the `BlockSpaceManagerImpl` instead of handling each case inside `Scheduler`.

DarkLight1337 commented on vllm/engine/llm_engine.py:
```diff
@@ -754,9 +757,17 @@ def has_unfinished_requests_for_virtual_engine(
+        model_mode: ModelMode,
     ) -> None:
-        seq_group.embeddings = outputs[0].embeddings
+        if model_mode == ModelMode.EMBEDDING:
+            seq_group.embeddings = outputs[0].embeddings
+        elif model_mode == ModelMode.SIMPLE:
+            seq_group.result = outputs[0].result
+        else:
+            raise ValueError(
+                "Only support ModelMode.EMBEDDING Or ModelMode.SIMPLE")
```
I wonder whether we could also factor out this logic into the individual `ModelMode`s....

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -285,7 +287,9 @@ def generate(
             instead pass them via the ``inputs`` parameter.
         """
-        if self.llm_engine.model_config.embedding_mode:
+        if self.llm_engine.model_config.model_mode in [
```
I think whitelisting the model modes would be easier to maintain instead of blacklisting them.

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -429,7 +433,8 @@ def encode(
         """
-        if not self.llm_engine.model_config.embedding_mode:
+
+        if self.llm_engine.model_config.model_mode != ModelMode.EMBEDDING:
```
`is not` check should work with enums.

DarkLight1337 commented on vllm/entrypoints/openai/serving_embedding.py:
```diff
@@ -136,8 +136,8 @@ async def create_embedding(self, request: EmbeddingRequest,
-    def _check_embedding_mode(self, embedding_mode: bool):
-        if not embedding_mode:
+    def _check_embedding_mode(self, model_mode: ModelMode):
+        if model_mode != ModelMode.EMBEDDING:
```
`is not` check should work with enums.

DarkLight1337 commented on vllm/worker/worker.py:
```diff
@@ -85,8 +86,10 @@ def __init__(
         if model_runner_cls is not None:
             ModelRunnerClass = model_runner_cls
-        elif self.model_config.embedding_mode:
+        elif self.model_config.model_mode == ModelMode.EMBEDDING:
```
Perhaps we should let each mode override the `ModelRunnerClass` instead of handling each case inside `Worker`.

AllenDou replied:
I made some changes as you can see
But I didn't quite understand your point about "override BlockSpaceManagerImpl ModelRunnerClass". Do you mean to override the entire BlockManager/ModelRunner class? That seems more complicated than the current implementation.

DarkLight1337 replied:
Here, by "override" I just mean set the `ModelRunnerClass` to use. We can add static methods to `ModelMode` (e.g. `get_block_space_manager_impl` and `get_model_runner_cls`) so we can keep the switch/case code all in one place.

AllenDou replied:
Please take a look class ModelMode, some 'circular import' problems occur, but solved.

DarkLight1337 commented on vllm/model_executor/models/__init__.py:
```diff
@@ -132,8 +183,23 @@ def register_model(model_arch: str, model_cls: Type[nn.Module]):
     @staticmethod
-    def is_embedding_model(model_arch: str) -> bool:
-        return model_arch in _EMBEDDING_MODELS
+    def get_model_mode(architectures: List) -> ModelMode:
```
```suggestion
    def get_model_mode(architectures: List[str]) -> ModelMode:
```

DarkLight1337 commented on vllm/model_executor/models/__init__.py:
```diff
@@ -132,8 +183,23 @@ def register_model(model_arch: str, model_cls: Type[nn.Module]):
+        return ModelMode.DECODER
+
+    @staticmethod
+    def need_initialize_kv_caches(modelmode: ModelMode) -> bool:
+        if modelmode == ModelMode.EMBEDDING:
+            return False
+        if modelmode == ModelMode.SIMPLE:
+            return False
+        return True
```
```suggestion
    def need_initialize_kv_caches(model_mode: ModelMode) -> bool:
        if model_mode == ModelMode.EMBEDDING:
            return False
        if model_mode == ModelMode.SIMPLE:
            return False
        return True
```

DarkLight1337 commented on vllm/model_executor/models/__init__.py:
```diff
@@ -73,7 +74,57 @@
+    SIMPLE = enum.auto()
+
+    @staticmethod
+    def get_model_runner_cls(model_mode):
```
```suggestion
    def get_model_runner_cls(model_mode: "ModelMode"):
```

DarkLight1337 commented on vllm/model_executor/models/__init__.py:
```diff
@@ -73,7 +74,57 @@
+            return SimpleModelRunner
+
+    @staticmethod
+    def get_block_space_manager_impl(use_v2_block_manager, model_mode):
```
```suggestion
    def get_block_space_manager_impl(use_v2_block_manager: bool, model_mode: "ModelMode"):
```

DarkLight1337 reviewed (commented):
Added some suggestions to improve the type annotations.

DarkLight1337 commented on vllm/config.py:
```diff
@@ -152,6 +153,7 @@ def __init__(
                                     code_revision, rope_scaling, rope_theta)
         self.hf_text_config = get_hf_text_config(self.hf_config)
         self.dtype = _get_and_verify_dtype(self.hf_text_config, dtype)
+        self.model_mode = ModelMode.DECODER
```
Maybe can remove `_verify_model_mode` and just set it here.

AllenDou replied:
done.

DarkLight1337 commented on vllm/config.py:
```diff
@@ -766,19 +767,24 @@ def __init__(self,
                 # and TTFT on A100. Note it is not optimized for throughput.
                 self.max_num_batched_tokens = 512
-            elif embedding_mode:
+            elif model_mode == ModelMode.EMBEDDING:
                 # For embedding, choose specific value for higher throughput
                 self.max_num_batched_tokens = max(
                     max_model_len, _EMBEDDING_MODEL_MAX_NUM_BATCHED_TOKENS)
+            elif model_mode == ModelMode.SIMPLE:
+                # For non-[decoder,embedding] model, choose specific value
+                # for higher throughput
+                self.max_num_batched_tokens = max(
+                    max_model_len, _SIMPLE_MODEL_MAX_NUM_BATCHED_TOKENS)
```
Can also factor this out to `ModelMode`. (Return `_EMBEDDING_MODEL_MAX_NUM_BATCHED_TOKENS` etc. inside `ModelMode`, but keep the `max(...)` logic here.)

AllenDou replied:
Please take a look :)

DarkLight1337 commented on vllm/config.py:
```diff
@@ -772,15 +770,10 @@ def __init__(self,
-                self.max_num_batched_tokens = max(
-                    max_model_len, _SIMPLE_MODEL_MAX_NUM_BATCHED_TOKENS)
+                    max_model_len,
+                    ModelMode.get_model_max_num_batched_tokens(model_mode))
```
How about we return `None` for the other model modes (so we don't need the `elif` statement for this)? Afterwards we only apply `max` if the model mode returns a valid integer.

AllenDou replied:
Only keep "if enable_chunked_prefill" "else"?   remove "elif:"?

DarkLight1337 replied:
Yeah. We can apply the `max` part inside the `else` block

AllenDou replied:
Sorry, didn't get your point.

DarkLight1337 commented on vllm/model_executor/models/__init__.py:
```diff
@@ -73,7 +74,72 @@
+        return BlockSpaceManagerV1
+
+    @staticmethod
+    def get_model_max_num_batched_tokens(model_mode: "ModelMode"):
```
```suggestion
    def get_model_max_num_batched_tokens(model_mode: "ModelMode") -> Optional[int]:
```

DarkLight1337 commented on vllm/core/interfaces.py:
```diff
@@ -40,6 +40,11 @@ def get_block_space_manager_class(version: str):
+            from vllm.core.simple_model_block_manager import (
+                SimpleModelBlockSpaceManager)
+            return SimpleModelBlockSpaceManager
+
```
Can we remove the "embedding" and "simple" versions and instead return the class directly inside `ModelMode`? (Only keep "v1" and "v2" since they are actually versions)

DarkLight1337 replied:
Edit: Actually I think this function isn't used anymore?

AllenDou replied:
removed.

DarkLight1337 commented on vllm/worker/worker.py:
```diff
@@ -82,11 +81,12 @@ def __init__(
-        ModelRunnerClass: Type[GPUModelRunnerBase] = ModelRunner
+        #ModelRunnerClass: Type[GPUModelRunnerBase] = ModelRunner
         if model_runner_cls is not None:
             ModelRunnerClass = model_runner_cls
-        elif self.model_config.embedding_mode:
-            ModelRunnerClass = EmbeddingModelRunner
+        else:
+            ModelRunnerClass = ModelMode.get_model_runner_cls(
+                self.model_config.model_mode)
```
I think we can simplify this to a one-liner

```py
ModelRunnerClass = model_runner_cls or ModelMode.get_model_runner_cls(self.model_config.model_mode)
```

AllenDou replied:
Yes, get_block_space_manager_class() is not used anymore, replaced by get_block_space_manager_impl()

DarkLight1337 reviewed (approved):
The config and multi-modal parts LGTM. I assume the model is implemented correctly since the tests pass.

However, since I'm not involved with the internals of model executor, block manager and workers, I'll leave it to @robertgshaw2-neuralmagic to review those. He will also see how to integrate this into the existing work for encoder-decoder support.

robertgshaw2-neuralmagic commented:
Im going to take a look at this over the weekend. Thanks @AllenDou!

AllenDou commented:
mark https://github.com/vllm-project/vllm/issues/6424

AllenDou commented:
/ready

AllenDou commented:
Hello @robertgshaw2-neuralmagic , just a friendly reminder to review this PR when you get a chance,
We want to enable more models through this feature,  thank a lot!

robertgshaw2-neuralmagic commented:
Thanks @AllenDou ! this is on my list

AllenDou commented:
mark https://github.com/vllm-project/vllm/issues/6789

AllenDou commented:
mark https://github.com/vllm-project/vllm/issues/1187 https://github.com/vllm-project/vllm/issues/205

DarkLight1337 commented:
I think @robertgshaw2-neuralmagic has been very busy lately... Are you still interested in implementing this? Let's split the PR into smaller parts that are more manageable so that @robertgshaw2-neuralmagic doesn't have to review everything at once.

(By the way, please take into account the changes from #4942.)

AllenDou commented:
> I think @robertgshaw2-neuralmagic has been very busy lately... Are you still interested in implementing this? Let's split the PR into smaller parts that are more manageable so that @robertgshaw2-neuralmagic doesn't have to review everything at once.
> 
> (By the way, please take into account the changes from #4942.)

Of course, after reviewing https://github.com/vllm-project/vllm/pull/4942, I realized that nm folks have their own methods for supporting 'anymodel'. So, I think maybe I should close this PR. If this feature is needed in the future, I can reopen this PR at any time.

robertgshaw2-neuralmagic commented:
Hey @AllenDou Im really sorry for the delay here, have had no bandwidth

#4942 is only for encoder-decoder, which is separate from the implementation here, so I think we should be able to include both

AllenDou commented:
> Hey @AllenDou Im really sorry for the delay here, have had no bandwidth over the past month
> 
> #4942 is only for encoder-decoder, which is separate from the feature here, so I do not think this work is a duplicate
> 
> I’m happy to pick this back up if you want to reopen the PR

Great, I'll resolve the conflicts and reopen the PR, I hope this won't take up too much of your time :)

robertgshaw2-neuralmagic commented:
Thanks so much, I really appreciate your patience. Resolving our performance issues has just become a huge priority

AllenDou commented:
wwww

AllenDou commented:
Hi @robertgshaw2-neuralmagic , I have rebased and reopened this PR. Please take a look at it. The offline_inference_xlmroberta_awq.py file is a quantized version I've been working on, but I'm encountering some technical difficulties that are proving challenging to overcome. So, this file could be removed if necessary. cc @DarkLight1337

AllenDou commented:
offline_inference_xlmroberta_awq.py is deleted, as after hacking autoawq for xlmroberta model, I see no gain of performance benifit under vllm serving. Trying to fp8.

Commit:
[Core][Model] Add simple_model_runner and a new model XLMRobertaForSequenceClassification through multimodal interface

Commit:
fix model_mode

Commit:
fix model_mode

prashil1996 commented:
@AllenDou @robertgshaw2-neuralmagic @DarkLight1337 
Is there any timeline on when this could be delivered?

AllenDou commented:
> @AllenDou @robertgshaw2-neuralmagic @DarkLight1337 Is there any timeline on when this could be delivered?

Maybe we should wait until @robertgshaw2-neuralmagic gets a chance to review this PR.

DarkLight1337 commented:
A quick heads-up that the new locations of the model tests have been adjusted in #7820, so please merge from main.

DarkLight1337 commented:
Also @robertgshaw2-neuralmagic do you have any timeframe on when you will be available to review this PR?

fan-niu commented:
@AllenDou Hello, thank you for your contribution. I use concurrency to request services, and there will be a series of error reports, including the following errors:
` File "/share5/projects/llm/finetune-accelerate/inference/tensorrtllm-experiment/sglang/code/vllm_test/vllm_sqclassification/vllm/vllm/model_executor/models/xlmroberta.py", line 721, in forward
    batch_size, seq_length = input_shape
ValueError: not enough values to unpack (expected 2, got 1)`

Can you please help to optimize async requests? Thanks

Note: 
I use  http://127.0.0.1:8081/v1/completions, concurrency = 2,
payload:
{
        "prompt": "what time is it",
        "model":MODEL_NAME,
        "stop": "<\s>",
        "max_tokens":128,
        "stream": False,
        "presence_penalty": 0,
        "frequency_penalty": 0,
        "temperature": 0.0,
        "top_p": 1,
    }

AllenDou commented:
> @AllenDou Hello, thank you for your contribution. I use concurrency to request services, and there will be a series of error reports, including the following errors: ` File "/share5/projects/llm/finetune-accelerate/inference/tensorrtllm-experiment/sglang/code/vllm_test/vllm_sqclassification/vllm/vllm/model_executor/models/xlmroberta.py", line 721, in forward batch_size, seq_length = input_shape ValueError: not enough values to unpack (expected 2, got 1)`
> 
> Can you please help to optimize async requests? Thanks
> 
> Note: I use http://127.0.0.1:8081/v1/completions, concurrency = 2, payload: { "prompt": "what time is it", "model":MODEL_NAME, "stop": "<\s>", "max_tokens":128, "stream": False, "presence_penalty": 0, "frequency_penalty": 0, "temperature": 0.0, "top_p": 1, }

This PR currently does not support frontend access through the http protocol. By the way, the XLM-Roberta model compares the similarity of two strings. Therefore, you need to pass a tuple containing two strings (string, string) as input. Please refer to examples/offline_inference_xlmroberta.py for more details.

fan-niu commented:
> [RFC]: Support encode only models by Workflow Defined Engine #8453

Thanks for your reply, can you tell me how to add support for http requests on your branch? Thank you, or can you add this function if it is convenient for you?

wangyuanxiong-hub commented:
> @AllenDou Thanks for your reply, can you tell me how to add support for http requests on your branch? Thank you, or can you add this function if it is convenient for you? My final goal is to need llamaForSequenceClassification based on http request, I have added llamaForSequenceClassification to llama.py with reference to your branch, and implemented the forward function, but currently I cannot make the correct request through http.

Is it accessible via http now?

wangyuanxiong-hub commented:
The test results were not as expected, and the input and output methods were very confusing.

AllenDou commented:
> > @AllenDou Thanks for your reply, can you tell me how to add support for http requests on your branch? Thank you, or can you add this function if it is convenient for you? My final goal is to need llamaForSequenceClassification based on http request, I have added llamaForSequenceClassification to llama.py with reference to your branch, and implemented the forward function, but currently I cannot make the correct request through http.
> 
> Is it accessible via http now?

I apologize for the delayed response. I've been working on cute&ptx recently, therefore, this PR will not be updated for now :(

DarkLight1337 commented:
Closing as superseded by #10400. Sorry your PR didn't make it!
